HW2/P3/mandelbrot.pyx

```diff
@@ -41,8 +43,57 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                         break
                     z = z * z + c
                 out_counts[i, j] = iter
+    '''
+    with nogil:
+        for i in prange(in_coords.shape[0], schedule="static", chunksize=1, num_threads=4):
+            for j in xrange(0,in_coords.shape[1],8):
+                avxcr = AVX.make_float8(in_coords[i, j+7].real,
+                                        in_coords[i, j+6].real,
+                                        in_coords[i, j+5].real,
+                                        in_coords[i, j+4].real,
+                                        in_coords[i, j+3].real,
+                                        in_coords[i, j+2].real,
+                                        in_coords[i, j+1].real,
+                                        in_coords[i, j+0].real)
+
+                avxci = AVX.make_float8(in_coords[i, j+7].imag,
+                                        in_coords[i, j+6].imag,
+                                        in_coords[i, j+5].imag,
+                                        in_coords[i, j+4].imag,
+                                        in_coords[i, j+3].imag,
+                                        in_coords[i, j+2].imag,
+                                        in_coords[i, j+1].imag,
+                                        in_coords[i, j+0].imag)
+
+                # Keep track of the real part and imaginary part.
+                avxzr = AVX.float_to_float8(0.0)
+                avxzi = AVX.float_to_float8(0.0)
+
+                # Keep track of the magnitude
+                avxmag = AVX.float_to_float8(0.0)
+
+                # Check if the magnitude is less than 4.0
+                avxmask = AVX.less_than(avxmag,AVX.float_to_float8(4.0))
+                avxiter = AVX.float_to_float8(0.0)
 
+                for iter in range(max_iterations):
+                    # Increment iteration
+                    avxiter = AVX.add(avxiter, AVX.bitwise_and(AVX.float_to_float8(1.0), avxmask))
+
+                    # break if all of the 8 values are greater than 8.
+                    if (AVX.signs(avxmask) == 0):
+                        break
 
+                    # Update the real part and the imaginary part.
+                    avxzr, avxzi = AVX.add(AVX.fmsub(avxzr, avxzr, AVX.mul(avxzi, avxzi)), avxcr), AVX.add(AVX.fmadd(avxzr, avxzi, AVX.mul(avxzr, avxzi)), avxci)
+              
+                    avxmag = AVX.fmadd(avxzr, avxzr, AVX.mul(avxzi, avxzi))
+                    avxmask = AVX.less_than(avxmag, AVX.float_to_float8(4.0))
+                    
+                # Copy them to memory
```

Comment:
I think you needed to write a helper function to avoid collisions across threads in assigning values to out_vals (or fix the memory structure, which I don't think I saw).